File: take_u_forward/graph/27_Word Ladder I.java
import java.util.*;

class Pair {
    String first;
    int second;

    Pair(String first, int second) {
        this.first = first;
        this.second = second;
    }
}

class Solution {
    public int wordLadderLength(String startWord, String targetWord,
                                String[] wordList) {

        Queue<Pair> q = new LinkedList<>();
        q.add(new Pair(startWord, 1));

        Set<String> set = new HashSet<>();
        for (String word : wordList) {
            set.add(word);
        }

        set.remove(startWord);

        while (!q.isEmpty()) {
            Pair p = q.poll();
            String word = p.first;
            int steps = p.second;

            if (word.equals(targetWord))
                return steps;

            char[] arr = word.toCharArray();

            for (int i = 0; i < arr.length; i++) {
                char original = arr[i];

                for (char ch = 'a'; ch <= 'z'; ch++) {
                    arr[i] = ch;
                    String newWord = new String(arr);

                    if (set.contains(newWord)) {
                        set.remove(newWord);
                        q.add(new Pair(newWord, steps + 1));
                    }
                }
                arr[i] = original;
            }
        }

        return 0;
    }
}

File: take_u_forward/graph/27_Word Ladder I.cpp
class Solution {
  public:
    int wordLadderLength(string startWord, string targetWord,
                         vector<string>& wordList) {

        // Queue for BFS: {currentWord, steps}
        queue<pair<string, int>> q;
        q.push({startWord, 1});

        // Store all words in a set for fast lookup
        unordered_set<string> st(wordList.begin(), wordList.end());

        // Remove startWord if present
        if (st.find(startWord) != st.end())
            st.erase(startWord);

        while (!q.empty()) {
            auto it = q.front();
            q.pop();

            string word = it.first;
            int steps = it.second;

            // If target reached
            if (word == targetWord)
                return steps;

            // Try changing each character
            for (int i = 0; i < word.size(); i++) {
                char original = word[i];

                for (char ch = 'a'; ch <= 'z'; ch++) {
                    word[i] = ch;

                    if (st.find(word) != st.end()) {
                        st.erase(word);
                        q.push({word, steps + 1});
                    }
                }

                // Restore original character
                word[i] = original;
            }
        }

        // No transformation sequence found
        return 0;
    }
};
